### ISSUE TITLE
Watchers are notified before patch queue has been updated

### ISSUE THREAD

**Issue by alpeware** [evidentsystems/converge#18]
Hey Bobby,

Thanks for making the library available! I've enjoyed your recent talk at the Clojure meetup and today's oNote webinar - very exciting stuff.

While playing with the library, I noticed that the watch functions are called in the apply-state! fn before the patch queue is updated.

I suggest moving the `notify-w` fn to the `reset` fn -

```
    (reset
       [this new-value]
       (let [patch     (-make-patch this new-value)
             new-state (-state-from-patch this patch)
             old-value (:value state)]
         (assert (= new-value (:value new-state)) "Unsupported reference state")
         (-apply-state! this new-state)
         (when patch (set! patches (conj patches patch)))
         (notify-w this watches old-value (:value new-state))
         (:value new-state)))
```

and

```
    (-reset!
       [this new-value]
       (let [patch     (-make-patch this new-value)
             new-state (-state-from-patch this patch)
             old-value (:value state)]
         (assert (= new-value (:value new-state)) "Unsupported reference state")
         (-apply-state! this new-state)
         (when patch (set! patches (conj patches patch)))
         (notify-w this watches old-value (:value new-state))
         (:value new-state)))
```

This would allow clearing the patch queue in the ref's watch fn.

Let me know if this is something you're interested in.

**Comment by bobby** [comment]
We need `-apply-state!` to invoke the watches for cases such as `merge!`, but I think we could reverse the order in which we update the patch queue and apply the state.  Would that suffice for your case?

**Comment by bobby** [comment]
Fixed by 59c2324

**Comment by alpeware** [comment]
Just tested it and it works as I'd expect. Thanks for the quick change!
